docs/source/modifying-messages.rst

```diff
@@ -0,0 +1,390 @@
+``pad_byte`` specifies the value of the added bytes.
+
+In other calling convention, it accepts literal bytes to add to payload using
+the ``pad`` keyword argument.
+
+For example, to add 10 bytes with value 0 at the end of ClientHello, write:
+
+.. code:: python
+
+    ciphers = [CipherSuite.TLS_RSA_WITH_AES_128_CBC_SHA]
+    exts = {ExtensionType.renegotiation_info: None}
+    msg_gen = ClientHelloGenerator(cipihers, extensions=exts)
+    node = node.add_child(pad_handshake(msg_gen, 10))
+
+You can find example use of it in
+`test-truncating-of-client-hello.py
+<https://github.com/tomato42/tlsfuzzer/blob/master/scripts/test-truncating-of-client-hello.py>`_
+script.
+
+If you want to remove bytes from the end of a message, you can either
+specify negative ``size`` or use the
+:py:func:`~tlsfuzzer.messages.truncate_handshake`.
+
+.. note::
+
+    Sender can format ClientHello in two ways: with and without extensions.
+    ClientHello with empty list of extensions differs from one without
+    extensions by two zero bytes (they encode the length of the extensions).
+    Thus adding 2 zero bytes to an extensions-less ClientHello or removing
+    enough bytes from a ClientHello with extensions to turn it into one
+    without extensions can cause the
+    :py:func:`~tlsfuzzer.messages.pad_handshake` to create a well-formed
+    message, despite modifying it.
+
+Modifying content
+-----------------
+
+The :py:func:`~tlsfuzzer.messages.fuzz_message` supports changing arbitrary
+parts of sent messages.
+
+Both optional parameters of the function, ``substitutions`` and ``xors`` expect
+a dictionary as a value.
+The keys of the dictionary specify the bytes to change.
+To specify the bytes counting from the end of the message use negative numbers.
+
+For example, to change type of a ClientHello message to that of ServerHello
+use the following code:
+
+.. code:: python
+
+    ciphers = [CipherSuite.TLS_RSA_WITH_AES_128_CBC_SHA]
+    exts = {ExtensionType.renegotiation_info: None}
+    msg_gen = ClientHelloGenerator(cipihers, extensions=exts)
+    node = node.add_child(fuzz_message(msg_gen,
+                                       {0: HandshakeType.server_hello}))
+
+You can find example use of it in
+`test-invalid-client-hello.py
+<https://github.com/tomato42/tlsfuzzer/blob/master/scripts/test-invalid-client-hello.py>`_
+script.
+
+Modifying records
+=================
+
+The :term:`TLS` protocol specifies four types of encrypted records:
+ones that use stream encryption, ones that use block encryption in
+:term:`MAC` then encrypt mode, ones that use block encryption in encrypt then
+:term:`MAC` mode, and ones that use :term:`AEAD` ciphers.
+Each of them behaves differently on record layer level, thus modifying the
+intermediate ciphertext requires use of different functions.
+
+Fuzzing the MAC
+---------------
+
+Depending on cipher suite and negotiated extensions, to change the
+authentication tag you need to use different functions.
```

Comment:
To change the authentication tag you need to use different functions depending on which cipher suite and extensions have been negotiated.

Reply:
done